src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,168 @@
+ * without requiring a local checkout.
+ */
+public class GitRepoSizeEstimator {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
+    public static final int SIZE_TO_SWITCH = 50000;
+
+    /**
+     * Instantiate class using {@link AbstractGitSCMSource}. It looks for a cached .git directory first, calculates the
+     * size if it is found else checks if the extension point has been implemented and asks for the size.
+     * @param source the {@link AbstractGitSCMSource}
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public GitRepoSizeEstimator(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        boolean useAPI = false;
+
+        useCache = setSizeFromCache(source);
+
+        if (useCache) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            useAPI = setSizeFromAPI(source.getRemote());
+        }
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+
+        if (!useAPI && !useCache) {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * Estimate size of a repository using the extension point
+     * @param remoteName: The URL of the repository
+     */
+    public GitRepoSizeEstimator(String remoteName) {
+        boolean useAPI = setSizeFromAPI(remoteName);
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * For a given recommended git implementation, validate if the installation exists and provide no suggestion if
+     * implementation doesn't exist.
+     * @param gitImplementation: The recommended git implementation, "git" or "jgit" on the basis of the heuristics.
+     */
+    private void determineGitTool(String gitImplementation) {
+        if (gitImplementation.equals("DEFAULT")) {
+            gitTool = "NONE";
+            return; // Recommend nothing (GitToolRecommendation = NONE)
+        }
+        final Jenkins jenkins = Jenkins.get();
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param source: Use a {@link AbstractGitSCMSource} to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean setSizeFromCache(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
```

Comment:
Not a blocker, just a matter of taste. Let's try to avoid branches in the code
```suggestion
        boolean useCache = false;
```